// illusion/illusion/illusion.js
var path = require('path');
var fs = require('fs');
var Phantom = require('./phantom');
var Task = require('./task');

function Illusion() {
    this.phantom = new Phantom();
}

Illusion.prototype.createTask = function(url) {
    return new Task(this.phantom, url);
};

Illusion.prototype.init = function(cookiesPath, next) {
    var self = this;
    if (typeof cookiesPath === 'function') {
        next = cookiesPath;
        cookiesPath = undefined;
    }
    this.phantom.open(function(err, result) {
        if (err) {
            return next && next(err);
        }
        if (cookiesPath) {
            try {
                var modPath = require.resolve('illusion');
                var relPath = path.relative(modPath, process.cwd());
                cookiesPath = path.normalize(relPath + '/' + cookiesPath);
            } catch (e) {}
            var cookies;
            try {
                cookies = require(cookiesPath);
            } catch (e) {
                console.log('Cookies file not exist');
                return next && next(null, result);
            }
            self.phantom.addCookies(cookies, next);
        } else {
            return next && next(null, result);
        }
    });
};

Illusion.prototype.cleanup = function(cookiesPath, next) {
    var self = this;
    if (typeof cookiesPath === 'function') {
        next = cookiesPath;
        cookiesPath = undefined;
    }
    if (cookiesPath) {
        this.saveCookies(cookiesPath, function(err) {
            if (err) {
                return next && next(err);
            }
            self.phantom.close(next);
        });
    } else {
        this.phantom.close(next);
    }
};

Illusion.prototype.saveCookies = function(cookiesPath, next) {
    var self = this;
    this.phantom.getCookies(function(err, cookies) {
        if (err) {
            return next && next(err);
        }
        fs.writeFile(cookiesPath + '.json', JSON.stringify(cookies, null, 4), next);
    });
};

module.exports = Illusion;
