// illusion/illusion/phantomrequest.js
var http = require('http');
var extend = require('util-extend');
var queryString = require('querystring');

function PhantomRequest(port, pageID) {
    this.options = {
        hostname: 'localhost',
        port: port,
        path: pageID ? '/' + pageID + '/' : '/',
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': 0
        }
    };
}

PhantomRequest.prototype.post = function(cmd, param, next) {
    var data = { 'command': cmd };
    if (typeof param === 'function') {
        next = param;
    } else {
        data = extend(data, param || {});
    }
    data = queryString.stringify(data);
    this.options.headers['Content-Length'] = data.length;
    var req = http.request(this.options, function(res) {
        res.setEncoding('utf-8');
        var buf = '';
        res.on('data', function(data) {
            buf += data;
        });
        res.on('end', function() {
            if (next) {
                var contentType = res.headers['content-type'];
                if (contentType == 'text/plain') {
                    next(null, buf);
                } else if (contentType == 'text/json') {
                    next(null, JSON.parse(buf));
                } else {
                    next(new Error('Uknown type'));
                }
            }
        });
    });
    req.on('error', function(err) {
        return next && next(err);
    });
    req.write(data);
    req.end();
};

module.exports = PhantomRequest;
